config/v1/types_infrastructure.go

```diff
@@ -620,9 +620,83 @@ const (
+
+	// GCPServiceEndpointNameFile is the name used for the GCP File Service endpoint.
+	GCPServiceEndpointNameFile GCPServiceEndpointName = "File"
+
+	// GCPServiceEndpointNameIAM is the name used for the GCP IAM Service endpoint.
+	GCPServiceEndpointNameIAM GCPServiceEndpointName = "IAM"
+
+	// GCPServiceEndpointNameServiceUsage is the name used for the GCP Service Usage Service endpoint.
+	GCPServiceEndpointNameServiceUsage GCPServiceEndpointName = "ServiceUsage"
+
+	// GCPServiceEndpointNameStorage is the name used for the GCP Storage Service endpoint.
+	GCPServiceEndpointNameStorage GCPServiceEndpointName = "Storage"
+
+	// GCPServiceEndpointNameTagManager is the name used for the GCP Tag Manager Service endpoint.
+	GCPServiceEndpointNameTagManager GCPServiceEndpointName = "TagManager"
+)
+
+// GCPServiceEndpoint store the configuration of a custom url to
+// override existing defaults of GCP Services.
+type GCPServiceEndpoint struct {
+	// name is the name of the GCP service whose endpoint is being overridden.
+	// This must be provided and cannot be empty.
+	//
+	// Allowed values are Compute, Container, CloudResourceManager, DNS, File, IAM, ServiceUsage,
+	// Storage, and TagManager.
+	//
+	// As an example, when setting the name to Compute all requests made by the caller to the GCP Compute
+	// Service will be directed to the endpoint specified in the url field.
+	//
+	// +required
+	Name GCPServiceEndpointName `json:"name"`
+
+	// url is a fully qualified URI that overrides the default endpoint for a client using the GCP service specified
+	// in the name field.
+	// url is required, must use the scheme https, must not be more than 253 characters in length,
+	// and must be a valid URL according to Go's net/url package (https://pkg.go.dev/net/url#URL)
+	//
+	// An example of a valid endpoint that overrides the Compute Service: "https://compute-myendpoint1.p.googleapis.com"
+	//
+	// +required
+	// +kubebuilder:validation:MaxLength=253
+	// +kubebuilder:validation:XValidation:rule="isURL(self)",message="must be a valid URL"
+	// +kubebuilder:validation:XValidation:rule="isURL(self) ? (url(self).getScheme() == \"https\") : true",message="scheme must be https"
+	URL string `json:"url"`
+}
+
 // GCPPlatformSpec holds the desired state of the Google Cloud Platform infrastructure provider.
 // This only includes fields that can be modified in the cluster.
-type GCPPlatformSpec struct{}
+type GCPPlatformSpec struct {
+	// serviceEndpoints specifies endpoints that override the default endpoints
+	// used when creating clients to interact with GCP services.
+	// serviceEndpoints is optional.
+	// Only 1 endpoint override is permitted for each GCP service.
+	// The maximum number of endpoint overrides allowed is 9.
+	// +listType=atomic
+	// +kubebuilder:validation:MaxItems=9
+	// +kubebuilder:validation:XValidation:rule="self.all(x, self.exists_one(y, x.name == y.name))",message="only 1 endpoint override is permitted for a GCP service"
```

Comment:
Maybe be explicit that it's the name that's unique?
```suggestion
	// +kubebuilder:validation:XValidation:rule="self.all(x, self.exists_one(y, x.name == y.name))",message="only 1 endpoint override is permitted per GCP service name"
```